### PR TITLE
adjust catkey form to use multiple rows

### LINKED ISSUE #3358: Update catkey form
See https://github.com/sul-dlss/cocina-models/pull/373

So that users can choose the "refresh" or "do not refresh" setting for any associated catkey.  See comments below with design.

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made? 🤔

Fixes #3358 - update catkey editing form so it allows for adding/removing catkeys and setting of refresh/no refresh.  Design in the ticket: https://github.com/sul-dlss/argo/issues/3358#issuecomment-1116329820

As implemented:

Screen Shot 2022-05-11 at 2 03 14 PM

Also: validate that there is no more than a single catkey set to refresh and that any entered catkey values are valid..

~~HOLD for PO testing~~

## How was this change tested? 🤨

- Localhost browser
- Updated existing tests for new UI
- Added many more tests for all of the scenarios (deleting, updating, setting refresh)

### CODE REVIEW

File: app/controllers/catkeys_controller.rb
Comment (peetucket) [inline app/controllers/catkeys_controller.rb:6]: no longer needed, this is being caught higher up the stack: https://github.com/sul-dlss/argo/blob/main/app/models/repository.rb#L26

File: app/controllers/catkeys_controller.rb
Comment (peetucket) [inline app/controllers/catkeys_controller.rb:10]: No longer needed, being caught further up the stack: https://github.com/sul-dlss/argo/blob/main/app/models/repository.rb#L26

File: app/services/catkey.rb
Comment (peetucket) [inline app/services/catkey.rb:4]: move these to the global constants so I can use them in the new code above

File: app/services/catkey.rb
Comment (peetucket) [inline app/services/catkey.rb:34]: Note: this class is no longer used for individual catkey editing as changed in this PR, but it is still being used in the change set persister service classes, so I am leaving as is. 

see https://github.com/sul-dlss/argo/blob/main/app/services/collection_change_set_persister.rb#L61 and https://github.com/sul-dlss/argo/blob/main/app/services/item_change_set_persister.rb#L96

File: spec/requests/set_catkey_spec.rb
Comment (peetucket) [inline spec/requests/set_catkey_spec.rb:100]: no trimming of whitespace anymore, it will simply not validate anymore if you do this and refuse to save

**Review by ndushay — APPROVED** [review]
I think someone more reform fluent should review, but I'll approve.  Don't forget to remove "HOLD" in the PR title before merging.

File: app/controllers/catkeys_controller.rb
Comment (jcoyne) [inline app/controllers/catkeys_controller.rb:74]: This should be done via prepopulate!

File: app/controllers/catkeys_controller.rb
Comment (jcoyne) [inline app/controllers/catkeys_controller.rb:24]: Can you move this into CatkeyForm?

File: app/controllers/catkeys_controller.rb
Comment (jcoyne) [inline app/controllers/catkeys_controller.rb:7]: Can you move this into CatkeyForm?

File: app/views/catkeys/_row.html.erb
Comment (jcoyne) [inline app/views/catkeys/_row.html.erb:15]: Why do we need these fields?  Catkey's don't have an id and it doesn't make sense to mark them for destruction, since we're just rewriting the whole set rather than deleting records.

**Review by jcoyne — CHANGES_REQUESTED** [review]
I think we should find some time to work on this together.

File: app/views/catkeys/_edit.html.erb
Comment (jcoyne) [inline app/views/catkeys/_edit.html.erb:14]: I don't think this is an appropriate controller, because nested-form depends on the child resources actually being resources (like with an identifier) rather than parts of a bigger record.

File: app/controllers/catkeys_controller.rb
Comment (peetucket) [reply app/controllers/catkeys_controller.rb:7]: Yes, will move (though we may want to consider doing the same thing in the `TagsController` for consistency.  I'll create a ticket for that.

File: app/controllers/catkeys_controller.rb
Comment (peetucket) [reply app/controllers/catkeys_controller.rb:24]: yes, same comment as below for ModelProxy

File: app/controllers/catkeys_controller.rb
Comment (peetucket) [reply app/controllers/catkeys_controller.rb:57]: This is called when you update the catkeys.  The logic for saving to the cocina model is in the CatkeyForm class.

File: app/controllers/catkeys_controller.rb
Comment (peetucket) [reply app/controllers/catkeys_controller.rb]: moved the blank catkey row logic to a prepopulator on the form

File: app/controllers/catkeys_controller.rb
Comment (jcoyne) [reply app/controllers/catkeys_controller.rb:22]: My question is more about, when would the format be "html" rather than "turbo_stream"?

File: app/forms/catkey_form.rb
Comment (jcoyne) [inline app/forms/catkey_form.rb:71]: Can this destroy check be moved to the populator?

File: app/forms/catkey_form.rb
Comment (jcoyne) [inline app/forms/catkey_form.rb:63]: This causes a second load, which should be unnecessary, since we already loaded this in the controller.

File: app/forms/catkey_form.rb
Comment (jcoyne) [inline app/forms/catkey_form.rb:13]: can this use `Coercion` to be a boolean?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/controllers/catkeys_controller.rb
Comment (peetucket) [reply app/controllers/catkeys_controller.rb]: Got it.  Yes, I don't think the `respond_to` and formats are needed, it's a redirect in one case (valid) and a turbo response in the other (invalid).  Removed.

**Review by jcoyne — APPROVED** [review]

**Comment by jcoyne** [comment]
Please do a squash merge on this.

**Pull request merged**
